odoo/models.py

```diff
@@ -2453,15 +2456,32 @@ def _inherits_join_calc(self, alias, fname, query):
         """
         # INVARIANT: alias is the SQL alias of model._table in query
         model, field = self, self._fields[fname]
-        while field.inherited:
-            # retrieve the parent model where field is inherited from
-            parent_model = self.env[field.related_field.model_name]
-            parent_fname = field.related[0]
-            # JOIN parent_model._table AS parent_alias ON alias.parent_fname = parent_alias.id
-            parent_alias = query.left_join(
-                alias, parent_fname, parent_model._table, 'id', parent_fname,
-            )
-            model, alias, field = parent_model, parent_alias, field.related_field
+        def process(model, alias, field, sudo):
+            while field.related and not field.store:
+                sudo = sudo or field.compute_sudo
+                last_field = field.related_field
+                for join_fname in field.related[:-1]:
+                    field = model._fields[join_fname]
+                    model, alias, field, sudo = process(model, alias, field, sudo)
```

Comment:
This looks complicated:
```
def proces(...):
    while ...:
        for ....:
            proces(...)
```

but I don't think there is a way to simplify it

Reply:
Indeed, and that would be a good reason to document it more. Especially the loops and recursive call. It's not clear what part of the problem each of them is handling.

What's the responsibility of the `while` loop compared to the `for` loop ?

What's the responsibility of the recursive call compared to the encompassing block ?

I'm having a hard time seeing when will the `while` loop do several laps, as it (erroneously) seem to me that the `for` loop and recursive call are doing everything.

Reply:
I'll try to make some drawing to explain. It's not clear enough for me too :D -- I made the tests first and then made this code almost by intuition

Reply:
A possible option to don't use both while-loop and for-loop is something like this;

```
def process(...):
    if field is not related:
        return
    for ....:
        process(...)
        if it's last element in for-loop:
           proces(...)
        else:
            do join
```

Not sure if it's easy to read -- we get two places to call process instead of one

---

I've added comment to current implementation. Please check if it's good enough now

Reply:
So, it's
```
def process(...):
    if field is not related:
        return
    for join_fname in field.related[:-1]::
        process(...)
        do join
    call process() for last field
```
:thinking: looks better :+1: 
I'll try

Reply:
Done! Thanks!